cypress/integration/login.ts

```diff
@@ -0,0 +1,20 @@
+import { providers } from 'ethers';
+
+import { TestProvider } from '../util';
+
+context('Coordinape', () => {
+  before(() => {
+    cy.on('window:before:load', win => {
+      const provider = new providers.Web3Provider(
+        new TestProvider('http://localhost:8545')
+      );
+      Object.defineProperty(win, 'ethereum', { value: provider });
+    });
+  });
+  it('can login', () => {
+    cy.visit('http://localhost:3000');
+    cy.get('button').click();
```

Comment:
I wonder if we should setup some "best practices" and guides for how to interact with our interface and app. I've sometimes thought this as:

- minimize the possibility of text changes breaking UI tests. For example, this test is less focused on the exact text/design of the login button, but more on the functionality of it. In this way, I've thought in the past that it is preferable to use HTML #id or .class tags for selecting elements, rather than doing text searching and 'contains" checks. In this way, if we have a .login element, we can change the design whenever we want, but also have a nice interface for referring to this element in our tests at a "semantic" level. Does this make sense? A pattern we should use?

We might need to add more id and class elements to our UI.

I know some folks think this is a bad pattern, and would love to learn more about the downsides of this approach as well. @levity can you add your thoughts on the negatives of this id/class pattern?

@topocount where are the guides that Cypress recommends on using id/class?

Reply:
@crabsinger here's the explainer. Turns out the don't recommend using id/class but rather custom attributes. https://docs.cypress.io/guides/references/best-practices#Selecting-Elements

Reply:
@crabsinger Here's React Testing Library's take:

> The React Testing Library is a very light-weight solution for testing React components. It provides light utility functions on top of react-dom and react-dom/test-utils, in a way that encourages better testing practices. Its primary guiding principle is:
>
> [The more your tests resemble the way your software is used, the more confidence they can give you.](https://testing-library.com/docs/guiding-principles)
>
> So rather than dealing with instances of rendered React components, your tests will work with actual DOM nodes. The utilities this library provides facilitate querying the DOM in the same way the user would. Finding form elements by their label text (just like a user would), finding links and buttons from their text (like a user would). It also exposes a recommended way to find elements by a data-testid as an "escape hatch" for elements where the text content and label do not make sense or is not practical.

Reply:
I like how Cypress and RTL suggest that if you do need to select a node by a non-human-readable characteristic, that it's a custom attribute (`data-cy` or `data-testid`) so you don't overload an attribute that already has a different purpose.